atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/MarkAndCasConsensusForgettingStore.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.atomic;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CheckAndSetException;
+import com.palantir.atlasdb.keyvalue.api.CheckAndSetRequest;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.MultiCheckAndSetException;
+import com.palantir.atlasdb.keyvalue.api.MultiCheckAndSetRequest;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.logsafe.Preconditions;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class MarkAndCasConsensusForgettingStore implements ConsensusForgettingStore {
+    private final byte[] inProgressMarker;
+    private final KeyValueService kvs;
+    private final TableReference tableRef;
+    private final ReadableConsensusForgettingStore reader;
+
+    public MarkAndCasConsensusForgettingStore(byte[] inProgressMarker, KeyValueService kvs, TableReference tableRef) {
+        Preconditions.checkArgument(!kvs.getCheckAndSetCompatibility().consistentOnFailure());
+        this.inProgressMarker = inProgressMarker;
+        this.kvs = kvs;
+        this.tableRef = tableRef;
+        this.reader = new ConsensusForgettingStoreReaderImpl(kvs, tableRef);
+    }
+
+    @Override
+    public void mark(Cell cell) {
+        mark(ImmutableSet.of(cell));
+    }
+
+    @Override
+    public void mark(Set<Cell> cells) {
+        // we mark at lowest possible non-negative timestamp to make marking visible to all transactions.
```

Comment:
I think this isn't actually the reason here, since the Reader reads at Long.MAX_VALUE anyway (notice that these values are non-transactional). The actual reason is because atomic updates generally happen at the Cassandra wall clock time of the operation, so we need the mark to have a lower writetime than what any atomic update might attempt to write